### PR TITLE
Add a subproject with a JUnit session listener

### PR DESCRIPTION
**Pull request by Matyrobbrt**
Will allow running JUnits in a bootstrapped Minecraft environment.

### CODE REVIEW

File: loader/src/main/java/net/neoforged/fml/loading/targets/JUnitDevLaunchTarget.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+package net.neoforged.fml.loading.targets;
+
+import net.neoforged.api.distmarker.Dist;
+
+public class JUnitDevLaunchTarget extends CommonDevLaunchHandler {
+    @Override
+    public Dist getDist() {
+        return Dist.DEDICATED_SERVER;
```
Comment (Technici4n) [inline loader/src/main/java/net/neoforged/fml/loading/targets/JUnitDevLaunchTarget.java:13]: We need a way to load client classes too.

File: loader/src/main/java/net/neoforged/fml/loading/targets/JUnitDevLaunchTarget.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    @Override
+    protected void runService(String[] arguments, ModuleLayer gameLayer) throws Throwable {
+        Class.forName(gameLayer.findModule("neoforge").orElseThrow(), "net.neoforged.neoforge.junit.JUnitMain").getMethod("main", String[].class).invoke(null, (Object)arguments);
+    }
+
+    @Override
+    public String name() {
+        return "junitfmldev";
```
Comment (Technici4n) [inline loader/src/main/java/net/neoforged/fml/loading/targets/JUnitDevLaunchTarget.java:23]: `forgejunitdev`?

File: loader/src/main/java/net/neoforged/fml/loading/targets/JUnitUserDevLaunchTarget.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    @Override
+    protected void runService(String[] arguments, ModuleLayer gameLayer) throws Throwable {
+        Class.forName(gameLayer.findModule("neoforge").orElseThrow(), "net.neoforged.neoforge.junit.JUnitMain").getMethod("main", String[].class).invoke(null, (Object)arguments);
+    }
+
+    @Override
+    public String name() {
+        return "junitfmluserdev";
```
Comment (Technici4n) [inline loader/src/main/java/net/neoforged/fml/loading/targets/JUnitUserDevLaunchTarget.java:23]: `forgejunituserdev`?

File: loader/src/main/java/net/neoforged/fml/loading/targets/JUnitDevLaunchTarget.java (hunk 1 above)
Comment (Matyrobbrt) [reply loader/src/main/java/net/neoforged/fml/loading/targets/JUnitDevLaunchTarget.java:13]: The dist of the launch target doesn't change the classes being loaded, just configures the RDC. Configuration which will be overwritten by the neo bootstrap.

**Comment by marchermans** [comment]
Outside of some documentation as to what the components do, this is good from my end.

File: loader/build.gradle [hunk 4]
```diff
@@ -36,5 +36,9 @@ dependencies {
 
     testCompileOnly("org.jetbrains:annotations:${jetbrains_annotations_version}")
     testRuntimeOnly("cpw.mods:bootstraplauncher:${bootstraplauncher_version}")
+
+    constraints {
```
[the remark this answers is not in the document]
Comment (Matyrobbrt) [reply loader/build.gradle:40]: It adds the junit version as a BOM entry, so one doesn't need to specify the version.

File: junit/src/main/java/module-info.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+import net.neoforged.fml.junit.JUnitService;
+import org.junit.platform.launcher.LauncherSessionListener;
+
+module net.neoforged.fancymodlauncher.junit {
```
Comment (Technici4n) [inline junit/src/main/java/module-info.java:9]: it's fancymod**loader**

Comment (Technici4n) [reply junit/src/main/java/module-info.java:9]: fml might be even better

File: junit/src/main/resources/META-INF/services/org.junit.platform.launcher.LauncherSessionListener [hunk 6]
```diff
@@ -0,0 +1 @@
+net.neoforged.fml.junit.JUnitService
```
Comment (Technici4n) [inline junit/src/main/resources/META-INF/services/org.junit.platform.launcher.LauncherSessionListener:1]: Any reason to specify the services both in the module-info and in META-INF? Do we expect users to launch this outside of the module path?

File: loader/build.gradle (hunk 4 above)
[the remark this answers is not in the document]
Comment (Technici4n) [reply loader/build.gradle:40]: Can you explain this stuff with a comment?

**Review by Technici4n — COMMENTED** [review]
I also want to see the NeoForge PR using this. ;)

**Review by marchermans — APPROVED** [review]

**Review by marchermans — APPROVED** [review]

**Pull request merged**